### DISCUSSION TITLE
How can I create a new instance of a Ray struct?

### DISCUSSION THREAD

**Discussion by Patrick-van-Halm** [vfsfitvnm/frida-il2cpp-bridge#336] [Q&A] [answered]
I want to call a method with a custom Ray object. A ray has a constructor of Vector3 Origin and Vector3 Direction, these are also structs.

**Comment by maarsalien** [comment, marked as answer]
I managed to do this, but i am myself still a little bit confused about how the ``unbox`` work maybe @vfsfitvnm  will have a good explanation for all of us

```ts
import 'frida-il2cpp-bridge';

const log = console.log.bind(console);

function Vector3(x = 0, y = 0, z = 0) {
  const Vector3 = Il2Cpp.domain.assembly('UnityEngine.CoreModule').image.class('UnityEngine.Vector3');
  const vector3 = Vector3.alloc();
  vector3.method('.ctor', 3).invoke(x, y, z);
  return vector3;
}

function Ray(origin = Vector3(), direction = Vector3(0, 0, 1)) {
  const Ray = Il2Cpp.domain.assembly('UnityEngine.CoreModule').image.class('UnityEngine.Ray');
  const ray = Ray.alloc();
  ray.method('.ctor', 2).invoke(origin.unbox(), direction.unbox());
  return ray;
}

Il2Cpp.perform(() => {
  const origin = Vector3();
  const direction = Vector3(10, 20, 30);

  const ray = Ray(origin, direction);

  log();
  log('ray', ray);
  log('GetPoint', ray.method('GetPoint', 1).invoke(10));
});
```
![Screenshot from 2023-07-15 13-33-28](https://github.com/vfsfitvnm/frida-il2cpp-bridge/assets/85195577/8feda8b7-31c0-41b9-ba89-d7206609ce95)

**Comment by maarsalien** [comment]
and also, instead of changing the ray reference, you could just call the ``set_origin`` and ``set_direction`` on the intercepted method to replace the coordinates.
